src/main/src/main/java/org/geoserver/catalog/CatalogBuilder.java

```diff
@@ -1077,13 +1035,88 @@ private CoverageInfo buildCoverageInternal(GridCoverage2DReader reader, String n
+
+        if (customParameters != null) {
+        	parameters.putAll(customParameters);
         }
 
-        return cinfo;
+        // make sure mosaics with many superimposed tiles won't blow up with 
+        // a "too many open files" exception
+        String maxAllowedTiles = ImageMosaicFormat.MAX_ALLOWED_TILES.getName().toString();
+        if (parameters.keySet().contains(maxAllowedTiles)) {
+            parameters.put(maxAllowedTiles, 1);
+        }
+
+        // Since the read sample image won't be greater than 5x5 pixels and we are limiting the
+        // number of granules to 1, we may do direct read instead of using JAI
+        String useJaiImageRead = ImageMosaicFormat.USE_JAI_IMAGEREAD.getName().toString();
+        if (parameters.keySet().contains(useJaiImageRead)) {
+            parameters.put(useJaiImageRead, false);
+        }
+
+        parameters.put(AbstractGridFormat.READ_GRIDGEOMETRY2D.getName().toString(), new GridGeometry2D(testRange, testEnvelope));
+
+        // try to read this coverage
+        final GridCoverage2D gc = reader.read(CoverageUtils.getParameters(readParams, parameters, true));
+        final GridSampleDimension[] sampleDimensions;
+        if (gc != null) {
+            // remove read grid geometry since it is request specific
+            parameters.remove(AbstractGridFormat.READ_GRIDGEOMETRY2D.getName().toString());
+            sampleDimensions = gc.getSampleDimensions();
+            /// dispose coverage 
+            gc.dispose(true);
+            if(gc.getRenderedImage() instanceof PlanarImage) {
+                ImageUtilities.disposePlanarImageChain((PlanarImage) gc.getRenderedImage());
+            }
+        } else {
+            final ImageLayout imageLayout = reader.getImageLayout();
+            if(imageLayout == null) {
+                throw new Exception("Unable to acquire test coverage and image layout for format:" + format.getName());
+            }
+            ColorModel cm = imageLayout.getColorModel(null);
+            if(cm == null) {
+                throw new Exception("Unable to acquire test coverage and color model for format:" + format.getName());
+            }
+            SampleModel sm = imageLayout.getSampleModel(null);
+            if(cm == null) {
+                throw new Exception("Unable to acquire test coverage and sample model for format:" + format.getName());
+            }
+            final int numBands = sm.getNumBands();
+            sampleDimensions = new GridSampleDimension[numBands];
+            // setting bands names.
+            for (int i = 0; i < numBands; i++) {
+                final ColorInterpretation colorInterpretation = TypeMap.getColorInterpretation(cm, i);
+                if (colorInterpretation == null)
+                    throw new IOException("Unrecognized sample dimension type for band number " + (i + 1));
+                sampleDimensions[i] = new GridSampleDimension(colorInterpretation.name());
+            }
```

Comment:
some "if(condition)" bad formatting and an "if(colorInterpretation == null) body " without enclosing bracket.